src/spec.rs

```diff
@@ -140,6 +144,30 @@ impl Zip64CentralDirectoryEndLocator {
     }
 }
 
+#[cfg(feature = "tokio")]
+impl Zip64CentralDirectoryEndLocator {
+    pub async fn parse<T>(reader: &mut T) -> ZipResult<Self>
+    where
+        T: AsyncRead + Unpin,
+    {
+        let magic = reader.read_u32_le().await?;
+        if magic != ZIP64_CENTRAL_DIRECTORY_END_LOCATOR_SIGNATURE {
+            return Err(ZipError::InvalidArchive(
+                "Invalid zip64 locator digital signature header",
+            ));
+        }
+        let disk_with_central_directory = reader.read_u32_le().await?;
+        let end_of_central_directory_offset = reader.read_u64_le().await?;
+        let number_of_disks = reader.read_u32_le().await?;
+
+        Ok(Self {
+            disk_with_central_directory,
+            end_of_central_directory_offset,
+            number_of_disks,
+        })
+    }
+}
```

Comment:
```suggestion
    macro_rules! parse {
      ($maybe_await:ident) => {
        let magic = maybe_await(reader.read_u32_le())?;
        if magic != ZIP64_CENTRAL_DIRECTORY_END_LOCATOR_SIGNATURE {
            return Err(ZipError::InvalidArchive(
                "Invalid zip64 locator digital signature header",
            ));
            let disk_with_central_directory = maybe_await(reader.read_u32_le())?;
            let end_of_central_directory_offset = maybe_await(reader.read_u64_le()?);
            let number_of_disks = maybe_await(reader.read_u32_le())?;

        Ok(Self {
            disk_with_central_directory,
            end_of_central_directory_offset,
            number_of_disks,
        })
      }
    }
    
    impl Zip64CentralDirectoryEndLocator {
        #[cfg(feature = "tokio")]
        pub async fn parse<T>(reader: &mut T) -> ZipResult<Self>
            where
                T: AsyncRead + Unpin,
            {
                use std::future::Future;
                async fn await<T>(operand: Future<T>) -> T {
                    T.await
                }
                parse!(await)
            }
            
           #[cfg(feature = "tokio")]
           pub fn parse<T>(reader: &mut T) -> ZipResult<Self>
               where
                   T: Read
           {
               parse!(std::convert::identity)
           }
```

Reply:
I've been brainstorming using macros for sharing code between async and sync code but i can't find a way to create a macro that can handle both .await calls and calls without .await without breaking the other, like in the code u just showed, is the goal of supporting async is to have a different code pathway for people to use or to just be a wrapper around current sync functions?

Reply:
Maybe we need an extension trait that's implemented for both Read and AsyncRead, and other for Write and AsyncWrite. For the compression and decompression themselves, I think the wrapper approach is probably adequate, since they're CPU-heavy.

Reply:
I can make a demo in another branch of using async as just a wrapper around sync, but my preferred aproach would actually be using async all io calls and trying to create shareable sync functions for non io operations.

Reply:
I still don't understand why we can't use `maybe_await` and `sync_or_await_async` functions/macros to copy code to both sync and async versions of each function.

Reply:
Sorry for the long break. There's two issues in trying to share code between async and sync functions using macros, the first is that as long as you're using async-await syntax calling async inside the macro would break sync code and not calling .await breaks async code and using .await outside async code block is not allowed.
``` rust
macro_rules! maybe_await {
     (maybe_async:expr) => {
              $maybe_async().await.unwrap();
     }
}

fn main() {
     #[cfg!(feature = "async")]
     let string = maybe_await!(async {read_to_string("/tmp/foo")}); // Code breaks due to main not being async
     #[cfg!(feature = "sync")]
     let string = maybe_await!(read_to_string("/tmp/foo")) // Code breaks to to .await being called on sync function
     println!("{}", string)
```
I don't see a way to solve both problems while sharing any code whatsoever, if u can have any idea on how to due this then i'm all ears.

Reply:
For tests, one fix might be to use a `maybe_block_on!` macro as well. But maybe this is more trouble than it's worth; let's wait and see how much duplicated code is left after rebasing against #93 and factoring out as much as possible.

Reply:
It occurs to me that one solution might involve 0th-order macros called `sync_defs!` and `async_defs!` that would provide different definitions of 1st-order macros such as `maybe_await!` and `maybe_block_on!` and `read_or_async_read!`, and then definitions of sync and async versions of a function would invoke their different 0th-order macros and then the same 2nd-order macro. Could that possibly work?